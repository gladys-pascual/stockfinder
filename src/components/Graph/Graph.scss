.stock-graph-wrapper {
  margin: 50px 0px 100px 0px;
  display: flex;
  justify-content: center;
  .stock-graph {
    width: 90vw;
    .stock-buttons-container {
      display: flex;
      justify-content: center;
      button {
        width: 50px;
        padding: 8px;
      }
      .active {
        color: pink;
      }
    }
    .graph-line-chart {
      height: 450px;
    }
  }
}
