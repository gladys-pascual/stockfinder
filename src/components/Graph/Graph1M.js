import PropTypes from "prop-types";
import { ResponsiveLine } from "@nivo/line";
import "./Graph.scss";
import dayjs from "dayjs";
import useWindowSize from "../../hooks/useWindowSize";

const Graph1M = ({ close, date }) => {
  const formatDate = date.map((date, i) => {
    return dayjs.unix(date).format("DD MMM");
  });

  function aggregatingData(c, d) {
    const data = c.map((cost, index) => {
      return {
        x: d[index],
        y: cost,
      };
    });

    return data;
  }

  const data = aggregatingData(close, formatDate);
  const aggregatedData = [
    {
      id: "stock",
      data,
    },
  ];

  const color = close[close.length - 1] > close[0] ? "green" : "red";

  const size = useWindowSize();
  const axisBottomFormat =
    size.width < 1000
      ? null
      : {
          orient: "bottom",
          tickSize: 5,
          tickPadding: 5,
          tickRotation: 45,
          legend: "Date",
          legendOffset: 55,
          legendPosition: "middle",
        };

  return (
    <section className="graph-line-chart">
      <ResponsiveLine
        data={aggregatedData}
        margin={{ top: 50, right: 80, bottom: 80, left: 50 }}
        xScale={{ type: "point" }}
        yScale={{
          type: "linear",
          min: "auto",
          max: "auto",
        }}
        yFormat="<-.2f"
        colors={[color]}
        axisBottom={axisBottomFormat}
        axisRight={{
          orient: "right",
          tickSize: 2,
          tickPadding: 2,
          tickRotation: 0,
          legend: "Price",
          legendOffset: +60,
          legendPosition: "middle",
          format: (value) =>
            `${Number(value).toLocaleString("en-IN", {
              minimumFractionDigits: 2,
            })}`,
        }}
        axisLeft={null}
        pointSize={4}
        pointColor={{ theme: "background" }}
        pointBorderWidth={2}
        pointBorderColor={{ from: "serieColor" }}
        pointLabelYOffset={-12}
        useMesh={true}
        crosshairType={"bottom-right"}
        animate={false}
      />
    </section>
  );
};

Graph1M.propTypes = {
  close: PropTypes.array.isRequired,
  date: PropTypes.array.isRequired,
};

export default Graph1M;
