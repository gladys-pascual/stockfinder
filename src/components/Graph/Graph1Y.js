import PropTypes from "prop-types";
import { ResponsiveLine } from "@nivo/bar";
import "./Graph.scss";

const Graph1Y = ({ close, date }) => {
  console.log(close);
  console.log(date);

  return (
    <section className="graph-line-chart">
      <h1>Graph 1 year</h1>
    </section>
  );
};

Graph1Y.propTypes = {
  close: PropTypes.array.isRequired,
  date: PropTypes.array.isRequired,
};

export default Graph1Y;
