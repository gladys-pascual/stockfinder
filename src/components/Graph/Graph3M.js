import PropTypes from "prop-types";
import { ResponsiveLine } from "@nivo/line";
import "./Graph.scss";
import dayjs from "dayjs";
import useWindowSize from "../../hooks/useWindowSize";

const Graph3M = ({ close, date }) => {
  const formatDate = date.map((date, i) => {
    return dayjs.unix(date).format("YYYY-MM-DD");
  });

  function aggregatingData(c, d) {
    const data = c.map((cost, index) => {
      return {
        x: d[index],
        y: cost,
      };
    });

    return data;
  }

  const data = aggregatingData(close, formatDate);
  const aggregatedData = [
    {
      id: "stock",
      data,
    },
  ];

  const color = close[close.length - 1] > close[0] ? "green" : "red";

  const size = useWindowSize();
  const axisBottomFormat =
    size.width < 1000
      ? null
      : {
          orient: "bottom",
          tickSize: 5,
          tickPadding: 5,
          tickRotation: 45,
          legend: "Date",
          legendOffset: 40,
          legendPosition: "middle",
          format: "%d %b %y",
          tickValues: "every 8 days",
          legend: "Date",
          legendOffset: 60,
        };

  return (
    <section className="graph-line-chart">
      <ResponsiveLine
        data={aggregatedData}
        margin={{ top: 50, right: 80, bottom: 80, left: 50 }}
        xScale={{
          type: "time",
          format: "%Y-%m-%d",
          useUTC: false,
          precision: "day",
        }}
        xFormat="time:%d %b %Y"
        yScale={{
          type: "linear",
          min: "auto",
          max: "auto",
        }}
        yFormat="<-$.2f"
        colors={[color]}
        axisBottom={axisBottomFormat}
        axisRight={{
          orient: "right",
          tickSize: 2,
          tickPadding: 2,
          tickRotation: 0,
          legend: "Price",
          legendOffset: +60,
          legendPosition: "middle",
          format: (value) =>
            `${Number(value).toLocaleString("en-IN", {
              minimumFractionDigits: 2,
            })}`,
        }}
        axisLeft={null}
        pointSize={4}
        pointColor={{ theme: "background" }}
        pointBorderWidth={2}
        pointBorderColor={{ from: "serieColor" }}
        pointLabelYOffset={-12}
        useMesh={true}
        crosshairType={"bottom-right"}
        animate={false}
      />
    </section>
  );
};

Graph3M.propTypes = {
  close: PropTypes.array.isRequired,
  date: PropTypes.array.isRequired,
};

export default Graph3M;
