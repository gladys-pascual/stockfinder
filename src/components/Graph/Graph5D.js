import PropTypes from "prop-types";
import { ResponsiveLine } from "@nivo/line";
import "./Graph.scss";
import dayjs from "dayjs";
import useWindowSize from "../../hooks/useWindowSize";

const Graph5D = ({ close, date }) => {
  const formatDate = date.map((date, i) => {
    return dayjs.unix(date).format("HH:MM DD MMM");
  });

  function aggregatingData(c, d) {
    const data = c.map((cost, index) => {
      return {
        x: d[index],
        y: cost,
      };
    });

    return data;
  }

  const data = aggregatingData(close, formatDate);
  const aggregatedData = [
    {
      id: "stock",
      data,
    },
  ];

  const color = close[close.length - 1] > close[0] ? "green" : "red";

  const size = useWindowSize();
  const axisBottomFormat =
    size.width < 1000
      ? null
      : {
          orient: "bottom",
          tickSize: 5,
          tickPadding: 5,
          tickRotation: 45,
          legend: "Date",
          legendOffset: 70,
          legendPosition: "middle",
        };

  return (
    <section className="graph-line-chart">
      <ResponsiveLine
        data={aggregatedData}
        margin={{ top: 50, right: 80, bottom: 80, left: 50 }}
        xScale={{ type: "point" }}
        yScale={{
          type: "linear",
          min: "auto",
          max: "auto",
        }}
        yFormat="<-.2f"
        colors={[color]}
        axisBottom={axisBottomFormat}
        axisRight={{
          orient: "right",
          tickSize: 2,
          tickPadding: 2,
          tickRotation: 0,
          legend: "Price",
          legendOffset: +60,
          legendPosition: "middle",
          format: (value) =>
            `${Number(value).toLocaleString("en-IN", {
              minimumFractionDigits: 2,
            })}`,
        }}
        axisLeft={null}
        pointSize={4}
        pointColor={{ theme: "background" }}
        pointBorderWidth={2}
        pointBorderColor={{ from: "serieColor" }}
        pointLabelYOffset={-12}
        useMesh={true}
        crosshairType={"bottom-right"}
        animate={false}
      />
    </section>
  );
};

Graph5D.propTypes = {
  close: PropTypes.array.isRequired,
  date: PropTypes.array.isRequired,
};

export default Graph5D;
