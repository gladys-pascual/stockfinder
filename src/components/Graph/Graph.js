import "./Graph.scss";
import axios from "axios";
import PropTypes from "prop-types";
import Graph5D from "./Graph5D";
import Graph1M from "./Graph1M";
import Graph3M from "./Graph3M";
import Graph1Y from "./Graph1Y";
import Graph5Y from "./Graph5Y";
import { useState, useEffect, useMemo } from "react";
import LoadingStockGraph from "../../components/Loading/LoadingStockGraph";
import NoInfo from "../../components/NoInfo/NoInfo";

const Graph = ({ symbol }) => {
  const [selectedButton, setSelectedButton] = useState(1);
  const [stockData, setStockData] = useState(null);
  const [multiplier, setMultiplier] = useState(5);
  const [resolution, setResolution] = useState(60);
  const [loading, setLoading] = useState(true);
  const [noInfo, setNoInfo] = useState(false);

  //Get stock prices
  const now = useMemo(() => {
    return Math.floor(Date.now() / 1000);
  }, []);

  const startDate = now - 86400 * multiplier;

  const startDateHandler = (multiplier, resolution) => {
    setMultiplier(multiplier);
    setResolution(resolution);
  };

  useEffect(() => {
    async function fetchStockData() {
      try {
        setLoading(true);
        const result = await axios.get(
          `https://finnhub.io/api/v1/stock/candle?symbol=${symbol}&resolution=${resolution}&from=${startDate}&to=${now}&token=${process.env.REACT_APP_API_KEY}`
        );
        setStockData(result.data);
        setLoading(false);
      } catch (err) {
        console.log("Error fetching and parsing data", err);
        setNoInfo(true);
        setLoading(false);
      }
    }
    fetchStockData();
  }, [symbol, resolution, startDate, now]);

  if (loading) {
    return <LoadingStockGraph />;
  }

  if (noInfo) {
    return <NoInfo />;
  }

  if (
    stockData.s === "no_data" ||
    stockData === { error: "You don't have access to this resource." }
  ) {
    console.log("There is no available price graph for this stock.");
    return null;
  }

  let graphChart;
  if (multiplier === 5) {
    graphChart = <Graph5D close={stockData.c} date={stockData.t} />;
  } else if (multiplier === 30) {
    graphChart = <Graph1M close={stockData.c} date={stockData.t} />;
  } else if (multiplier === 180) {
    graphChart = <Graph3M close={stockData.c} date={stockData.t} />;
  } else if (multiplier === 360) {
    graphChart = <Graph1Y close={stockData.c} date={stockData.t} />;
  } else if (multiplier === 1800) {
    graphChart = <Graph5Y close={stockData.c} date={stockData.t} />;
  }

  const handleClick = (buttonNumber) => {
    setSelectedButton(buttonNumber);
  };

  return (
    <section className="stock-graph-wrapper">
      <div className="stock-graph">
        <div className="stock-buttons-container">
          <button
            className={selectedButton === 1 ? "active" : "not-active"}
            onClick={() => {
              startDateHandler(5, "60");
              handleClick(1);
            }}
          >
            5D
          </button>
          <button
            className={selectedButton === 2 ? "active" : "not-active"}
            onClick={() => {
              startDateHandler(30, "D");
              handleClick(2);
            }}
          >
            1M
          </button>
          <button
            className={selectedButton === 3 ? "active" : "not-active"}
            onClick={() => {
              startDateHandler(180, "D");
              handleClick(3);
            }}
          >
            3M
          </button>
          <button
            className={selectedButton === 4 ? "active" : "not-active"}
            onClick={() => {
              startDateHandler(360, "W");
              handleClick(4);
            }}
          >
            1Y
          </button>
          <button
            className={selectedButton === 5 ? "active" : "not-active"}
            onClick={() => {
              startDateHandler(1800, "M");
              handleClick(5);
            }}
          >
            5Y
          </button>
        </div>
        <div>{graphChart}</div>
      </div>
    </section>
  );
};

Graph.propTypes = {
  symbol: PropTypes.string.isRequired,
};

export default Graph;
