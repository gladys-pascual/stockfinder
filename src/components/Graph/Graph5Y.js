import PropTypes from "prop-types";
import { ResponsiveLine } from "@nivo/bar";
import "./Graph.scss";

const Graph5Y = ({ close, date }) => {
  console.log(close);
  console.log(date);

  return (
    <section className="graph-line-chart">
      <h1>Graph 5 years</h1>
    </section>
  );
};

Graph5Y.propTypes = {
  close: PropTypes.array.isRequired,
  date: PropTypes.array.isRequired,
};

export default Graph5Y;
