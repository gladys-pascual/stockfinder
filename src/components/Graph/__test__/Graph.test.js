import { render } from "@testing-library/react";
import Graph from "../Graph";

test("render Graph component as expected", () => {
  const props = {
    symbol: "AAPL",
  };

  const { container } = render(<Graph {...props} />);

  expect(container).toMatchSnapshot();
});
