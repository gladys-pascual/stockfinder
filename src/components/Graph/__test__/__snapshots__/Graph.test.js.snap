// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render Graph component as expected 1`] = `
<div>
  <div
    class="loading-stock-graph-header"
  >
    <h1
      class="loading-heading"
    >
      Loading your stock prices...
    </h1>
    <div
      class="loading-stock-graph"
    >
      <span
        class="css-8pkz"
      >
        <span
          class="css-rxlor6"
        />
        <span
          class="css-13u5b26"
        />
        <span
          class="css-144dyr7"
        />
        <span
          class="css-pu3zi4"
        />
        <span
          class="css-tzx4d9"
        />
        <span
          class="css-1mjpbtg"
        />
      </span>
    </div>
  </div>
</div>
`;
